src/main/java/org/rumbledb/compiler/VariableDependenciesVisitor.java

```diff
@@ -0,0 +1,466 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Authors: Stefan Irimescu, Can Berker Cikis
+ *
+ */
+
+package org.rumbledb.compiler;
+
+import org.jgrapht.graph.DefaultEdge;
+import org.jgrapht.graph.DirectedAcyclicGraph;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.Map;
+
+import org.rumbledb.config.RumbleRuntimeConfiguration;
+import org.rumbledb.exceptions.OurBadException;
+import org.rumbledb.exceptions.VariableAlreadyExistsException;
+import org.rumbledb.expressions.AbstractNodeVisitor;
+import org.rumbledb.expressions.Expression;
+import org.rumbledb.expressions.Node;
+import org.rumbledb.expressions.control.TypeSwitchExpression;
+import org.rumbledb.expressions.control.TypeswitchCase;
+import org.rumbledb.expressions.flowr.CountClause;
+import org.rumbledb.expressions.flowr.ForClause;
+import org.rumbledb.expressions.flowr.GroupByClause;
+import org.rumbledb.expressions.flowr.GroupByVariableDeclaration;
+import org.rumbledb.expressions.flowr.LetClause;
+import org.rumbledb.expressions.flowr.OrderByClause;
+import org.rumbledb.expressions.flowr.OrderByClauseSortingKey;
+import org.rumbledb.expressions.flowr.ReturnClause;
+import org.rumbledb.expressions.flowr.SimpleMapExpression;
+import org.rumbledb.expressions.flowr.WhereClause;
+import org.rumbledb.expressions.module.FunctionDeclaration;
+import org.rumbledb.expressions.module.Prolog;
+import org.rumbledb.expressions.module.VariableDeclaration;
+import org.rumbledb.expressions.postfix.DynamicFunctionCallExpression;
+import org.rumbledb.expressions.postfix.PredicateExpression;
+import org.rumbledb.expressions.primary.ContextItemExpression;
+import org.rumbledb.expressions.primary.FunctionCallExpression;
+import org.rumbledb.expressions.primary.InlineFunctionExpression;
+import org.rumbledb.expressions.primary.NamedFunctionReferenceExpression;
+import org.rumbledb.expressions.primary.VariableReferenceExpression;
+import org.rumbledb.expressions.quantifiers.QuantifiedExpression;
+import org.rumbledb.expressions.quantifiers.QuantifiedExpressionVar;
+
+
+/**
+ * Dynamic context visitor. Populates the dynamic context to evaluate the main expression.
+ */
+public class VariableDependenciesVisitor extends AbstractNodeVisitor<Void> {
+
+    private RumbleRuntimeConfiguration rumbleRuntimeConfiguration;
+
+    Map<Node, Set<String>> inputVariableDependencies;
+    Map<Node, Set<String>> outputVariableDependencies;
+
+    private void addInputVariableDependencies(Node node, Set<String> variables) {
+        if (variables == null) {
+            addInputVariableDependencies(node, Collections.emptySet());
+            return;
+        }
```

Comment:
This if block does not align with what the function name describes. I would expect this function to merely add variables dependencies list. However, if this function is every called with variables set as null, all previously stored set variable dependencies would be wiped and replaced with an emptySet(). This may cause bugs in the future.

Furthermore, I am sure that you are aware as well of the redundancy of creating sets (even as they are empty) when no dependencies exist. I am thinking of an alternative that uses getters returning a static empty set if the dependencies do not exist. This would avoid the redundantly generated sets for each node without variable dependencies.